### PR TITLE
Refactor `BuildPlan`, better document and explain `--config`.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/156591.
Closes https://github.com/flutter/flutter/issues/156355.

I also moved some functions only used as implementation details :)

### CODE REVIEW

File: tools/engine_tool/lib/src/build_plan.dart [hunk 1]
```diff
@@ -162,28 +163,36 @@ final class BuildPlan {
...
-      defaultsTo: () {
-        if (builds.any((b) => b.name == 'host_debug')) {
-          return 'host_debug';
-        }
-        return null;
-      }(),
+      help: ''
+          'Selects a build configuration for the current platform.\n'
+          '\n'
+          'If omitted attempts '
```
Comment (gaaclarke) [inline tools/engine_tool/lib/src/build_plan.dart:181]:
```suggestion
          'If omitted, et attempts '
```

File: tools/engine_tool/lib/src/build_plan.dart [hunk 2]
```diff
@@ -162,28 +163,36 @@ final class BuildPlan {
...
+          'to default to a suitable target platform, which is typically a '
```
Comment (gaaclarke) [inline tools/engine_tool/lib/src/build_plan.dart:182]:
```suggestion
          'to default to a suitable target platform. This is typically a '
```

File: tools/engine_tool/lib/src/build_plan.dart [hunk 3]
```diff
@@ -361,3 +374,41 @@ enum BuildStrategy {
...
+_ConfigsByName _runnableBuildConfigs(
+  Platform platform, {
+  required Map<String, BuilderConfig> configsByName,
+}) {
+  return configsByName.entries.where((entry) {
+    return entry.value.canRunOn(platform);
+  });
+}
+
+/// Creates a final listing of builds from a mapping of builds.
```
Comment (gaaclarke) [inline tools/engine_tool/lib/src/build_plan.dart:390]: FWIW it isn't clear what this does from here.  I'd say `Extracts [Build]s from [runnableConfigs] that can execute on [platform].

**Review by gaaclarke — APPROVED** [review]
lgtm, just a few notes on documentation

File: tools/engine_tool/lib/src/build_plan.dart (hunk 3 above)
Comment (matanlurey) [reply tools/engine_tool/lib/src/build_plan.dart]: Done!

**Pull request merged**